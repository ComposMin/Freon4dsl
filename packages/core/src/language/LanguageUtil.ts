import { Language } from "../storage/index";
import { PiBinaryExpression } from "./PiBinaryExpression";
import { PiElement } from "./PiElement";
import { PiExpression } from "./PiExpression";
import { PiModel } from "./PiModel";
import { PiModelUnit } from "./PiModelUnit";

export function isPiModel(element: PiElement): element is PiModel {
    return !!element && element.piIsModel && element.piIsModel();
}

export function isPiExpression(element: PiElement): element is PiExpression {
    return !!element && element.piIsExpression && element.piIsExpression();
}

export function isPiBinaryExpression(element: PiElement): element is PiBinaryExpression {
    return !!element && element.piIsExpression && element.piIsExpression() && element.piIsBinaryExpression && element.piIsBinaryExpression();
}

export function ownerOfType(element: PiElement, typename: string, exact?: boolean): PiElement {
    let parent = element.piOwnerDescriptor()?.owner;
    while (!!parent) {
        if (exact ? instanceOfExact(parent, typename) : instanceOfSub(parent, typename)) {
            return parent;
        } else {
            parent = parent.piOwnerDescriptor()?.owner;
        }
    }
    return null;
}

/**
 * Returns the modelunit that owns this `element'
 * @param element
 */
export function modelUnit(element: PiElement): PiModelUnit {
    let current = element;
    while (!!current) {
        if (current.piIsUnit()) {
            return current as PiModelUnit;
        } else {
            current = current.piOwnerDescriptor()?.owner;
        }
    }
    // No modelunit found, element is standalone
    return null;
}

export function instanceOfExact(src: PiElement, target: string): boolean {
    return src.piLanguageConcept() === target;
}

export function instanceOfSub(src: PiElement, target: string): boolean {
    return isSubConcept(src.piLanguageConcept(), target);
}

export function isExactConcept(src, target: string): boolean {
    return src === target;
}

export function isSubConcept(src, target: string): boolean {
    return src === target || Language.getInstance().classifier(target).subConceptNames.includes(src);
}

export function matchElementList(list: PiElement[], toBeMatched: Partial<PiElement>[]): boolean {
    let foundMatch: boolean = true;
    for (const theirs of toBeMatched) {
        let xx: boolean = false;
        for (const mine of list) {
            if (mine.match(theirs)) {
                xx = true;
                break;
            }
        }
        foundMatch = foundMatch && xx;
        if (!foundMatch) {
            return false;
        }
    }
    return foundMatch;
}
