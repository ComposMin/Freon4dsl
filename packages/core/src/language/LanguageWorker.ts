/**
 * Interface LanguageWorker defines the extended visitor pattern of instances of languages.
 * Every node is visited twice, once before the visit of its children, and once after this visit.
 */
import { PiElement } from "./PiElement";

export interface LanguageWorker {
    execBefore(modelelement: PiElement): boolean;
    execAfter(modelelement: PiElement): boolean;

}
