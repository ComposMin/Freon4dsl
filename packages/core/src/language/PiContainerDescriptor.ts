import { PiElement } from "./PiElement";

export interface PiContainerDescriptor {
    container: PiElement;
    propertyName: string;
    propertyIndex?: number;
}
