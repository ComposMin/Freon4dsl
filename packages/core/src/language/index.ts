export * from "./PiElement";
export * from "./PiNamedElement";
export * from "./PiExpression";
export * from "./PiBinaryExpression";
export * from "./PiContainerDescriptor";
export * from "./PiExpression";
export * from "./PiModel";
export * from "./decorators";
export * from "./LanguageUtil";
