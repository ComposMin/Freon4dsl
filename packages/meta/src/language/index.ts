export * from "./MetaGenerator";
export * from "./MetaModel";
export * from "./PiMetaTemplates";
export * from "./TestGenerator";
