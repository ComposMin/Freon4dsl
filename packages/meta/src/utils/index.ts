export * from "./Checker";
export * from "./Helpers";
export * from "./Names";
export * from "./PathProvider";
export * from "./PiParser";
export * from "./ModelHelpers";
