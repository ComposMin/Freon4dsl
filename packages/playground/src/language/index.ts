// Generated by the ProjectIt Language Generator.
export * from "./DemoAbsExpression";
export * from "./DemoAndExpression";
export * from "./DemoAttribute";
export * from "./DemoAttributeType";
export * from "./DemoBinaryExpression";
export * from "./DemoBooleanLiteralExpression";
export * from "./DemoComparisonExpression";
export * from "./DemoDivideExpression";
export * from "./DemoEntity";
export * from "./DemoEqualsExpression";
export * from "./DemoExpression";
export * from "./DemoFunction";
export * from "./DemoFunctionCallExpression";
export * from "./DemoGreaterThenExpression";
export * from "./DemoIfExpression";
export * from "./DemoLessThenExpression";
export * from "./DemoLiteralExpression";
export * from "./DemoModel";
export * from "./DemoMultiplyExpression";
export * from "./DemoNumberLiteralExpression";
export * from "./DemoOrExpression";
export * from "./DemoPlaceholderExpression";
export * from "./DemoPlusExpression";
export * from "./DemoStringLiteralExpression";
export * from "./DemoType";
export * from "./DemoVariable";
export * from "./DemoVariableRef";
export * from "./WithType";
