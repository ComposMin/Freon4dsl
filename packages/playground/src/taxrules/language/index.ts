// Generated by the ProjectIt Language Generator.
export * from "./AbsExpression";
export * from "./AllTaxRulesConcepts";
export * from "./AndExpression";
export * from "./BinaryExpression";
export * from "./BooleanLiteralExpression";
export * from "./ComparisonExpression";
export * from "./DivideExpression";
export * from "./EqualsExpression";
export * from "./Expression";
export * from "./GreaterThenExpression";
export * from "./IfExpression";
export * from "./IncomePart";
export * from "./IncomeType";
export * from "./LessThenExpression";
export * from "./LiteralExpression";
export * from "./Money";
export * from "./MoneyLiteralExp";
export * from "./MultiplyExpression";
export * from "./NumberLiteralExpression";
export * from "./OrExpression";
export * from "./PercentageExpression";
export * from "./PlaceholderExpression";
export * from "./PlusExpression";
export * from "./RevenueService";
export * from "./RuleParameter";
export * from "./StringLiteralExpression";
export * from "./TaxID";
export * from "./TaxPayer";
export * from "./TaxPayerType";
export * from "./TaxRule";
export * from "./TaxRuleCall";
export * from "./TaxRuleSet";
