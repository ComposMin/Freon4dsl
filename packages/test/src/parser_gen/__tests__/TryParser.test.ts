import { Demo, DemoUnit } from "../language/gen";
import { DemoEnvironment } from "../environment/gen/DemoEnvironment";
import { FileHandler } from "../../utils/FileHandler";

describe("Test the parser", () => {
    test( ": read two units and create a model", () => {
        const reader = DemoEnvironment.getInstance().reader;
        const writer = DemoEnvironment.getInstance().writer;
        const fileHandler = new FileHandler();

        let input = fileHandler.stringFromFile("src/parser_gen/__tests__/ParserInput1.txt");
        const unit1: DemoUnit = reader.readFromString(input, "DemoUnit") as DemoUnit;

        unit1.main?.baseEntity.forEach(ent => {
            expect(ent).not.toBeUndefined();
            expect(ent).not.toBeNull();
        });
        unit1.main?.attributes.forEach(attr => {
            expect(attr.declaredType).not.toBeUndefined();
            expect(attr.declaredType).not.toBeNull();
        });

        input = fileHandler.stringFromFile("src/parser_gen/__tests__/ParserInput2.txt");
        const unit2: DemoUnit = reader.readFromString(input, "DemoUnit") as DemoUnit;
        //
        unit2.main?.baseEntity.forEach(ent => {
            expect(ent).not.toBeUndefined();
            expect(ent).not.toBeNull();
        });
        unit2.main?.attributes.forEach(attr => {
            expect(attr.declaredType).not.toBeUndefined();
            expect(attr.declaredType).not.toBeNull();
        });
        //
        const myModels: DemoUnit[] = [];
        myModels.push(unit1);
        myModels.push(unit2);
        const model = Demo.create({ name: "SOME_MODEL", models: myModels });

        let output = writer.writeToString(model, 0, false);
        fileHandler.stringToFile("src/parser_gen/__tests__/ParserOutput.txt", output);
        expect(output).toMatchSnapshot();
    });
});
