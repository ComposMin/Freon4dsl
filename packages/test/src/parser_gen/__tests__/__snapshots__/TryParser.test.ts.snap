// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the parser : read two units and create a model 1`] = `
"unit eenNaam { 
main: entity1 base anotherEntity, 
                   andAnother, 
                   andYetAnother
      start 
          attr1 : type1  = \\"myString\\", 
          attr2 : type2  = \\"TEKST\\"
      end 
unit wide functions: 
    DemoFunction functionXXX 
        expression CALL funct555 (  ) 
        parameters 
            DemoVariable variableYYY 
                declaredType entity1
        declaredType STRING 
}"
`;

exports[`Test the parser : read two units and create a model 2`] = `
"unit andereNaam { 
main: entity56 base anotherEntity, 
                    andAnother, 
                    andYetAnother
      start 
          attr78 : type1  = \\"myString\\", 
          attr9 : type2  = \\"TEKST\\"
      end 
unit wide functions: 
    DemoFunction functionYYY 
        expression CALL funct555 (  ) 
        parameters 
            DemoVariable variableZZZ 
                declaredType entity56
        declaredType STRING 
}"
`;
